### DISCUSSION TITLE
Line graph of temperature in HASPmota

### DISCUSSION THREAD

**Discussion by richardsg307** [arendst/Tasmota#20348] [Q&A]
I've got a Lilygo T-Display-S3 and am using it to display various things like temperature as text labels using HASPmota.
I have a second page with some additional stuff but what I'd really like is to be able to display on that page (or a third) a graph showing temperature over the last 24Hrs.
Is that at all possible? Is there a Widget for it?
I see there might be a way using LVGL but would that work on a page created/defined in HASPmota?

**Comment by s-hadinger** [comment]
There is a good example in Berry-LVGL code using bar here: https://github.com/arendst/Tasmota/blob/development/tasmota/berry/haspmota_src/haspmota_widgets/lv_wifi_graph.be

Option 1 would be to create a widget in Berry and just use it in HASPMota.

Option 2 would be to have a `chart` object in HASPMota but I haven't mapped the `chart` LVGL to any HASPmota type. It seems that OpenHASP hasn't either.

Would option 1 work for you?

**Reply by richardsg307** [reply]
I guess that code is for the WiFi graph I seen but haven't tried. I'll try to work out how I might adapt it and where I'm supposed to put the code and how to call it!
Is there a guide on "how to create a Berry Widget for use in HASPmota?
Edit:
I've found the file haspmota.widgets.tapp and loaded it to my ESP32 File system. Also put the example code to call up the WiFi graph etc on my page 2 in my pages.jsonl file. Works OK.
But when /I look at the tapp file in Notepad++ i see strange bits that look like inappropriate rendering of machine code. What's that about?
This is the section before ```class lv_tasmota_info : lv.label``` (It looks a bit different here from how Notepad++ renders it)

```
PK
     nŠ†T B·‚   ‚     autoexec.beUT	 ¯Mb¯Mbux ø     # pre-load widgets so future `import` will be already in memory
import lv_tasmota_log
import lv_tasmota_info
import lv_wifi_graph
PK
     ñ°…Tjïj¹Ö  Ö    lv_tasmota_info.beUT	 †¡Lb¯Mbux ø     # lv_tasmota_info class
```
And not only that, but when I look at the file in the ESP32 File Management system big chunks are missing! eg the first 27 lines as per Notepad++ are replaced with ```PK```

**Reply by s-hadinger** [reply]
A tapp file is just a zip file uncompressed. You are not supposed to open them as text files. https://tasmota.github.io/docs/Tasmota-Application/

If you want a custom class in HASPmota, you just need to create the classe before starting HASPmota and reference it with its class name, like `{"id":20,"obj":"lv_wifi_graph","x":257,"y":25,"w":60,"h":40,"radius":0}`

I'm considering adding a simple chart for which you can easily add values with a cron. Hold on

**Reply by sfromis** [reply]
You can also find the HASPmota widgets files here, instead of screwing it up with Notepad++: https://github.com/arendst/Tasmota/tree/development/tasmota/berry/haspmota_src/haspmota_widgets

**Reply by s-hadinger** [reply]
I'm working on a chart widget. Stay tuned

**Reply by richardsg307** [reply]
Me too, but I'm sure you will do it much better and more generic. I've got mine plotting the Outflow and Inflow temperatures with 1200 points, one taken every 15 seconds giving 5 hours worth of data. I've also put grid lines on it but can't work out how to make them paler ie 20% opacity. Mapping from the LVGL website documentation to Tasmota implementation is a struggle. Also tried axis labels but got the impression they were not implemented.

**Reply by s-hadinger** [reply]
Please have a look at https://github.com/arendst/Tasmota/pull/20372

Documentation is not updated yet but main parts are in the demo

**Reply by sfromis** [reply]
https://tasmota.github.io/docs/HASPmota/#attributes-specific-to-chart
